### ISSUE TITLE
Remove 64 character limit on the mailbox check for validateAddress()

### ISSUE THREAD

**Issue by r-a-y** [PHPMailer/PHPMailer#706]
Hi,

When validating an email address, the `validateAddress()` method imposes a 64 character limit on the mailbox of an email address.

If an email address has a mailbox length that is larger than 64 characters, then `validateAddress()` will return false.

For example, the following would fail:

    SUPERLONGSTRINGSUPERLONGSTRINGSUPERLONGSTRINGSUPERLONGSTRINGSUPERLONGSTRING@example.com

`validateAddress()` claims to use the same spec as PHP's `FILTER_VALIDATE_EMAIL` filter.  However, `FILTER_VALIDATE_EMAIL` is based on the RFC 822 spec:
http://php.net/manual/en/filter.filters.validate.php

[Section 3.4.8 of the RFC 822 spec](https://www.w3.org/Protocols/rfc822/3_Lexical.html#z3) states:

>  For readability, the field-body portion of long header fields may be "folded" onto multiple lines of the actual field. "Long" is commonly interpreted to mean greater than 65 or 72 characters. The former length serves as a limit, when the message is to be viewed on most simple terminals which use simple display software; however, **the limit is not imposed by this standard** (emphasis added).

Therefore, I propose that the 64 character limit check be removed to be in compliance with the RFC 822 spec.

You can also test the `SUPERLONGSTRING` email address on this RFC 822 validation site:
http://sphinx.mythic-beasts.com/~pdw/cgi-bin/emailvalidate

And it passes as valid.

Also, if Github used PHPMailer for their emails, then their `Reply-To` email address would fail as their mailbox length is around 90 characters! :)

I'm currently working around using the `addAddress()` / `addCC()` / `addBCC()` / `addReplyTo()` methods by clearing out the email address field I want to set and redeclaring the header with the `addCustomHeader()` instead since there are no restrictions with that method.

**Comment by Synchro** [comment]
Section 3.4.8 describes general limits on the length of a message header, not the local part of an email address.

I'm very familiar with that checker - it was written by my ISP. RFC822 is not the only RFC that applies to addresses, for example it allows line breaks in addresses, which are not permitted in RFC821. Also, ICANN has ruled that 'dotless' domains are not permitted, so `a@b` is also considered invalid.

The local part is limited to 64 bytes in RFC 3696. There isn't an explicit limit on the local part in 822, but there is an implicit one - there is a limit on the overall length of an address (320 chars), and on the length of a domain part (256 chars), leaving only 64 available for the local part. Of course you may be lucky enough to have a domain shorter than 256 chars that allows you to have a longer local part in an address, however, since we're talking about RFC compliance, restricting to 64 is the only way to be absolutely sure. You can always override the validator function if you don't agree with it - PHPMailer 6.0 makes this easier by calling it as `static::validateAddress` from internal methods so that you can override it in subclasses.

PHP's `FILTER_VALIDATE_EMAIL` has a large number of holes - if you run the PHPMailer tests using that you'll find it fails on quite a few test cases, either allowing invalid addresses, or denying valid ones. That's why we have gone to the effort of having a more complex test. You might like to know that the PHP validator is based on an earlier version of the pattern that PHPMailer uses, written by the same author.

**Comment by r-a-y** [comment]
Thanks for the detailed reply.  I get that PHPMailer wants to be compliant.

Maybe update the DocBlock to remove the `FILTER_VALIDATE_EMAIL` mention as that was the cause of my confusion:
https://github.com/PHPMailer/PHPMailer/blob/v5.2.14/class.phpmailer.php#L1072

As I mentioned above, I'm already working around this behavior, but thought it was worth opening an issue to talk about it.  Feel free to close this one.

Also, #573 is a pretty good ticket as well so devs could set the default `$patternselect` instead of having to override the static validateAddress() method in v6.0.

**Comment by Synchro** [comment]
Incidentally:
```php
(boolean)var_dump(filter_var('1234567890123456789012345678901234567890123456789012345678901234@iana.org', FILTER_VALIDATE_EMAIL));
(boolean)var_dump(filter_var('12345678901234567890123456789012345678901234567890123456789012345@iana.org', FILTER_VALIDATE_EMAIL));
```
Gives:
```
bool(true)
bool(false)
```
i.e. FILTER_VALIDATE_EMAIL also rejects addresses with a local part longer than 64 chars. 

I've implemented a fix for #573 for the next release - actually going one better and allowing you to inject a callable to implement your own validator.
